File: javaCreateAccountCore/src/main/java/com/hamdan/forzenbook/java/createaccount/core/domain/usecase/CreateAccountUseCaseImpl.java
package com.hamdan.forzenbook.java.createaccount.core.domain.usecase;

import androidx.annotation.NonNull;

import com.hamdan.forzenbook.java.createaccount.core.data.repository.AccountException;
import com.hamdan.forzenbook.java.createaccount.core.data.repository.CreateAccountRepository;

public class CreateAccountUseCaseImpl implements CreateAccountUseCase {

    final private CreateAccountRepository repository;

    public CreateAccountUseCaseImpl(CreateAccountRepository repository) {
        this.repository = repository;
    }

    @Override
    @NonNull
    public CreateAccountResult invoke(String firstName, String lastName, String birthDay, String email, String location) {
        try {
            repository.createUser(firstName, lastName, birthDay, email, location);
            return CreateAccountResult.CREATE_SUCCESS;
        } catch (AccountException e) {
            return CreateAccountResult.CREATE_EXISTS;
        } catch (Exception e) {
            return CreateAccountResult.CREATE_FAILURE;
        }
    }
}

File: createAccountCore/src/main/java/com/hamdan/forzenbook/createaccount/core/domain/CreateAccountUseCaseImpl.kt
package com.hamdan.forzenbook.createaccount.core.domain

import android.util.Log
import com.hamdan.forzenbook.core.AccountException
import com.hamdan.forzenbook.createaccount.core.data.repository.CreateAccountRepository

class CreateAccountUseCaseImpl(
    val repository: CreateAccountRepository
) : CreateAccountUseCase {
    /**
     * returns a success, exists, or failure, failure indicates a critical error
     */
    override suspend fun invoke(
        firstName: String,
        lastName: String,
        birthDay: String,
        email: String,
        location: String
    ): CreateAccountResult {
        return try {
            repository.createUser(firstName, lastName, birthDay, email, location)
            CreateAccountResult.CREATE_SUCCESS
        } catch (e: AccountException) {
            Log.v("Exception", e.stackTraceToString())
            CreateAccountResult.CREATE_EXISTS
        } catch (e: Exception) {
            Log.v("Exception", e.stackTraceToString())
            CreateAccountResult.CREATE_FAILURE
        }
    }
}
